File: apps/agent/src/practices/eveningGratitude.ts
/**
 * Evening gratitude practice.
 *
 * When ON, the coach gently invites one thing the user is grateful for —
 * but only between 18:00 and 23:00 local time, and only if today's
 * entry hasn't been logged yet. The `log_gratitude` tool captures the
 * answer to user.yaml under `practices.evening_gratitude.entries[]`
 * and stamps `last_logged: <today YYYY-MM-DD>` so we don't double-ask.
 */

import { FunctionTool } from '@google/adk';
import { z } from 'zod';
import type { Practice, PracticeCtx, PracticeDeps } from './types.js';

const ID = 'evening_gratitude';
const ENTRIES_PATH = `practices.${ID}.entries`;
const LAST_LOGGED_PATH = `practices.${ID}.last_logged`;
const EVENING_START_HOUR = 18;
const EVENING_END_HOUR = 23;

interface GratitudeEntry {
  date: string; // YYYY-MM-DD local
  text: string;
  ts: string; // ISO UTC
}

function localDateAndHour(now: Date, tz: string | null): { date: string; hour: number } {
  // en-CA reliably returns YYYY-MM-DD in the chosen timezone; sv-SE for
  // 24-hour HH so we can read the hour without locale surprises.
  const tzOpt = tz ?? 'UTC';
  const date = new Intl.DateTimeFormat('en-CA', {
    timeZone: tzOpt,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  }).format(now);
  const hourStr = new Intl.DateTimeFormat('sv-SE', {
    timeZone: tzOpt,
    hour: '2-digit',
    hour12: false,
  }).format(now);
  return { date, hour: Number.parseInt(hourStr, 10) };
}

function directive(ctx: PracticeCtx): string | null {
  const { date: today, hour } = localDateAndHour(ctx.now, ctx.timezone);
  if (hour < EVENING_START_HOUR || hour >= EVENING_END_HOUR + 1) return null;
  const lastLogged =
    typeof ctx.practiceState.last_logged === 'string'
      ? (ctx.practiceState.last_logged as string)
      : null;
  if (lastLogged === today) return null;
  return `EVENING_GRATITUDE (practice on, evening window, not yet logged today):
It's evening and the user hasn't shared a gratitude entry yet today. When the moment fits — after a check-in, between topics, or as the chat winds down — gently invite one thing they're grateful for. ONE soft ask only; if they decline or change subject, drop it. When they share, immediately call log_gratitude({ text: "<their words, lightly cleaned up>" }) and continue normally without announcing the save.`;
}

function createLogGratitudeTool(deps: PracticeDeps, uid: string): FunctionTool {
  const parameters = z.object({
    text: z
      .string()
      .min(2)
      .describe("The user's gratitude in their own words. One short line is fine."),
  });

  // biome-ignore lint/suspicious/noExplicitAny: zod nominal mismatch with ADK generics
  return new FunctionTool<any>({
    name: 'log_gratitude',
    description:
      "Save the user's evening gratitude entry to their profile. Call this immediately " +
      "when the user shares one thing they're grateful for during an evening check-in. " +
      "Stores the text + today's date and stamps last_logged so we don't re-ask today. " +
      'Never announce the save.',
    parameters,
    execute: async (input: unknown) => {
      const { text } = input as { text: string };
      try {
        const now = new Date();
        // Use UTC for the date stamp at the storage layer; the directive
        // already enforces the evening window in user-local time, so a
        // UTC date here is fine for "did we log today?" idempotency.
        const date = now.toISOString().slice(0, 10);
        const ts = now.toISOString();
        const profile = await deps.profileStore.read(uid).catch(() => undefined);
        const existing = readEntries(profile);
        const next: GratitudeEntry[] = [...existing, { date, text, ts }];
        await deps.profileStore.updatePath(uid, ENTRIES_PATH, next);
        await deps.profileStore.updatePath(uid, LAST_LOGGED_PATH, date);
        return { status: 'ok' as const, count: next.length };
      } catch (err) {
        return {
          status: 'error' as const,
          message: err instanceof Error ? err.message : String(err),
        };
      }
    },
  });
}

function readEntries(profile: unknown): GratitudeEntry[] {
  if (!profile || typeof profile !== 'object') return [];
  const practices = (profile as Record<string, unknown>).practices;
  if (!practices || typeof practices !== 'object') return [];
  const slot = (practices as Record<string, unknown>)[ID];
  if (!slot || typeof slot !== 'object') return [];
  const entries = (slot as Record<string, unknown>).entries;
  if (!Array.isArray(entries)) return [];
  return entries.filter(
    (e): e is GratitudeEntry =>
      !!e &&
      typeof e === 'object' &&
      typeof (e as GratitudeEntry).date === 'string' &&
      typeof (e as GratitudeEntry).text === 'string' &&
      typeof (e as GratitudeEntry).ts === 'string',
  );
}

export const eveningGratitude: Practice = {
  id: ID,
  label: 'Evening gratitude',
  description: 'Each evening, the coach gently invites one thing you’re grateful for and saves it.',
  offerHint:
    'If the user mentions ending their day, winding down, or expresses positive reflection, consider offering Evening gratitude.',
  directive,
  tools: createLogGratitudeTool ? (deps, uid) => [createLogGratitudeTool(deps, uid)] : undefined,
};

File: apps/agent_py/src/lifecoach_agent/practices/evening_gratitude.py
"""Evening gratitude practice.

Mirrors `apps/agent/src/practices/eveningGratitude.ts`. Time gate
18:00–22:59 local, idempotent via `practices.evening_gratitude.last_logged`.
The `log_gratitude` tool factory is wired in Phase 6 once the
`UserProfileStore` lands; for Phase 3 only the directive is exposed.
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any

from lifecoach_agent.practices.day_clock import local_date_and_hour
from lifecoach_agent.practices.types import (
    Practice,
    PracticeCtx,
    PracticeDeps,
    PracticeTool,
)

ID = "evening_gratitude"
ENTRIES_PATH = f"practices.{ID}.entries"
LAST_LOGGED_PATH = f"practices.{ID}.last_logged"
EVENING_START_HOUR = 18
EVENING_END_HOUR = 23  # exclusive — 23:00 onwards skips


def directive(ctx: PracticeCtx) -> str | None:
    today, hour = local_date_and_hour(ctx.base.now, ctx.base.timezone)
    if hour < EVENING_START_HOUR or hour >= EVENING_END_HOUR:
        return None
    last_logged = ctx.practice_state.get("last_logged")
    if isinstance(last_logged, str) and last_logged == today:
        return None
    return (
        "EVENING_GRATITUDE (practice on, evening window, not yet logged today):\n"
        "It's evening and the user hasn't shared a gratitude entry yet today. When the "
        "moment fits — after a check-in, between topics, or as the chat winds down — "
        "gently invite one thing they're grateful for. ONE soft ask only; if they "
        "decline or change subject, drop it. When they share, immediately call "
        'log_gratitude({ text: "<their words, lightly cleaned up>" }) and continue '
        "normally without announcing the save."
    )


def _read_entries(profile: Any) -> list[dict[str, str]]:
    if not isinstance(profile, dict):
        return []
    practices = profile.get("practices")
    if not isinstance(practices, dict):
        return []
    slot = practices.get(ID)
    if not isinstance(slot, dict):
        return []
    entries = slot.get("entries")
    if not isinstance(entries, list):
        return []
    out: list[dict[str, str]] = []
    for e in entries:
        if (
            isinstance(e, dict)
            and isinstance(e.get("date"), str)
            and isinstance(e.get("text"), str)
            and isinstance(e.get("ts"), str)
        ):
            out.append(e)
    return out


def make_log_gratitude(deps: PracticeDeps, uid: str) -> PracticeTool:
    """Build the `log_gratitude(text=...)` callable. Captures `deps`/`uid`
    in the closure; the agent factory wraps the returned function in a
    FunctionTool with the right metadata."""

    async def log_gratitude(text: str) -> dict[str, Any]:
        try:
            now = datetime.now(UTC)
            date = now.strftime("%Y-%m-%d")
            ts = now.isoformat().replace("+00:00", "Z")
            try:
                profile = await deps.profile_store.read(uid)
            except Exception:
                profile = None
            existing = _read_entries(profile)
            entry = {"date": date, "text": text, "ts": ts}
            next_entries = [*existing, entry]
            await deps.profile_store.update_path(uid, ENTRIES_PATH, next_entries)
            await deps.profile_store.update_path(uid, LAST_LOGGED_PATH, date)
            return {"status": "ok", "count": len(next_entries)}
        except Exception as err:  # noqa: BLE001 — surface the message to the LLM
            return {"status": "error", "message": str(err)}

    return log_gratitude


def tools(deps: PracticeDeps, uid: str) -> list[PracticeTool]:
    return [make_log_gratitude(deps, uid)]


evening_gratitude = Practice(
    id=ID,
    label="Evening gratitude",
    description=(
        "Each evening, the coach gently invites one thing you’re grateful for and saves it."
    ),
    offer_hint=(
        "If the user mentions ending their day, winding down, or expresses positive "
        "reflection, consider offering Evening gratitude."
    ),
    directive=directive,
    tools=tools,
)
